pkgs/native_assets_builder/lib/src/dependencies_hash_file/dependencies_hash_file.dart

```diff
@@ -0,0 +1,211 @@
+    DateTime? validBeforeLastModified,
+  }) async {
+    _reset();
+
+    Uri? modifiedAfterTimeStamp;
+    for (final uri in fileSystemEntities) {
+      int hash;
+      if (validBeforeLastModified != null &&
+          (await uri.fileSystemEntity.lastModified())
+              .isAfter(validBeforeLastModified)) {
+        hash = _hashLastModifiedAfterCutoff;
+        modifiedAfterTimeStamp = uri;
+      } else {
+        if (_isDirectoryPath(uri.path)) {
+          hash = await _hashDirectory(uri);
+        } else {
+          hash = await _hashFile(uri);
+        }
+      }
+      _hashes.files.add(FilesystemEntityHash(uri, hash));
+    }
+    await _persist();
+    return modifiedAfterTimeStamp;
+  }
+
+  Future<void> _persist() => _file.writeAsString(json.encode(_hashes.toJson()));
+
+  /// Reads the file with hashes and finds an outdated file or directory if it
+  /// exists.
+  Future<Uri?> findOutdatedFileSystemEntity() async {
+    await _readFile();
+
+    for (final savedHash in _hashes.files) {
+      final uri = savedHash.path;
+      final savedHashValue = savedHash.hash;
+      final int hashValue;
+      if (_isDirectoryPath(uri.path)) {
+        hashValue = await _hashDirectory(uri);
+      } else {
+        hashValue = await _hashFile(uri);
+      }
+      if (savedHashValue != hashValue) {
+        return uri;
+      }
+    }
+    return null;
+  }
+
+  // A 64 bit hash from an md5 hash.
+  int _md5int64(Uint8List bytes) {
+    final md5bytes = md5.convert(bytes);
+    final md5ints = (md5bytes.bytes as Uint8List).buffer.asUint64List();
+    return md5ints[0];
+  }
+
+  Future<int> _hashFile(Uri uri) async {
+    final file = File.fromUri(uri);
+    if (!await file.exists()) {
+      return _hashNotExists;
+    }
+    return _md5int64(await file.readAsBytes());
+  }
+
+  Future<int> _hashDirectory(Uri uri) async {
+    final directory = Directory.fromUri(uri);
+    if (!await directory.exists()) {
+      return _hashNotExists;
+    }
+    final children = directory.listSync(followLinks: true, recursive: false);
+    final childrenNames = children.map((e) => _pathBaseName(e.path)).join(';');
+    return _md5int64(utf8.encode(childrenNames));
+  }
+
+  /// Predefined hash for files and directories that do not exist.
+  ///
+  /// There are two predefined hash values. The chance that a predefined hash
+  /// collides with a real hash is 2/2^64.
+  static const _hashNotExists = 0;
+
+  /// Predefined hash for files and directories that were modified after the
+  /// time that the hashes file was created.
+  ///
+  /// There are two predefined hash values. The chance that a predefined hash
+  /// collides with a real hash is 2/2^64.
+  static const _hashLastModifiedAfterCutoff = 1;
+}
+
+/// Storage format for file system entity hashes.
+///
+/// [File] hashes are a hash of the file.
+///
+/// [Directory] hashes are a hash of the names of the direct children.
+class FileSystemHashes {
+  FileSystemHashes({
+    this.version = 1,
```

Comment:
Why do we need a version?

In the current PR if this field every changed to a backwards incompatible version, it seems nothing would change because the code doesn't use this field at all (apart from serializing and deserializing it)